### PR TITLE
Show flag summary information to group moderators

### PR DESCRIPTION
**Pull request by fatbusinessman**
When presenting annotations for someone who is a moderator in their group, this will now include a `moderation` key in the annotation JSON. Currently it just shows the number of times each annotation has been flagged for the moderator’s attention.

My current plan:

- [x] Pull out the model logic into a service, akin to #4475 – this could go in a subsequent PR
- [x] Give the commits some hefty rejigging
- [x] Consider changing the name of the formatter class, since it could now be confused with the `AnnotationModeration` model
- [x] Decide whether to keep the functional test, remove it, or add to it

This work is for hypothesis/product-backlog#208.

### CODE REVIEW

File: h/formatters/annotation_moderation.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        self._cache.update(flag_counts)
+
+        return flag_counts
+
+    def format(self, annotation_resource):
+        if not self._has_permission('admin', annotation_resource.group):
+            return {}
+
+        flag_count = self._load(annotation_resource.annotation.id)
+        return {'moderation': {'flag_count': flag_count}}
```
Comment (nickstenning) [inline h/formatters/annotation_moderation.py]: Small thing, but I think we'll probably want `flagCount` here. We don't have a lot of multiword keys in the API output, the one exception being the selectors for the annotation, but those use `lowerCamelCase` rather than `lower_underscored`.

**Comment by nickstenning** [comment]
This is looking pretty good to me! FWIW, my feedback on your to-do list:

- services good ✔️ 
- hefty rejigging good ✔️ 
- meh, don't worry too much about the name -- we have package namespaces and [they're cool](https://www.python.org/dev/peps/pep-0020/)
- perhaps we could work the functional test into a more generic functional test of the annotation API in totality? (i.e. prepare a couple of annotations with various data, and check the output is as we expect, rather than checking each feature individually)

**Comment by nickstenning** [comment]
Oh, and I think having to pass `request.has_permission` around is a little unfortunate, but I can't immediately think of a better idea, so... ¯\_(ツ)_/¯

File: h/formatters/annotation_moderation.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        self._user = user
+        self._has_permission = has_permission
+
+        # Local cache of flag counts. We don't need to care about detached
+        # instances because we only store the annotation id and a count.
+        self._cache = {}
+
+    def preload(self, ids):
+        if self._user is None:
+            return
```
Comment (nickstenning) [inline h/formatters/annotation_moderation.py]: We probably also want an `if not ids` shortcut here, too.

File: h/formatters/annotation_moderation.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+        query = self._session.query(sa.func.count(models.Flag.id).label('flag_count'),
+                                    models.Flag.annotation_id) \
+                             .filter(models.Flag.annotation_id.in_(ids)) \
+                             .group_by(models.Flag.annotation_id)
```
Comment (nickstenning) [inline h/formatters/annotation_moderation.py]: This is very much a take it or leave it comment, but I tend to think that using parens for multi-line queries is a bit neater than escaping newlines:

```python
        query = (self._session.query(sa.func.count(models.Flag.id).label('flag_count'),
                                     models.Flag.annotation_id)
                 .filter(models.Flag.annotation_id.in_(ids))
                 .group_by(models.Flag.annotation_id))
```

File: h/formatters/annotation_moderation.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+        flag_counts = {f.annotation_id: f.flag_count for f in query}
+        missing_ids = set(ids) - set(flag_counts.keys())
```
Comment (nickstenning) [inline h/formatters/annotation_moderation.py]: Again, certainly not a blocker for merging this, but it might be slightly less work to do this the other way round. That is:

```python
flag_counts = {id_: 0 for id_ in ids}
flag_counts.update({f.annotation_id: f.flag_count for f in query})
```

File: h/formatters/annotation_moderation.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    def format(self, annotation_resource):
+        if not self._has_permission('admin', annotation_resource.group):
+            return {}
+
+        flag_count = self._load(annotation_resource.annotation.id)
+        return {'moderation': {'flagCount': flag_count}}
+
+    def _load(self, id_):
+        if self._user is None:
+            return False
```
Comment (nickstenning) [inline h/formatters/annotation_moderation.py]: It would be good to add a test for the negative branch here.

File: h/formatters/annotation_moderation.py [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+        if id_ in self._cache:
+            return self._cache[id_]
```
Comment (nickstenning) [inline h/formatters/annotation_moderation.py]: Ditto -- we should cover the positive branch here.

File: tests/h/formatters/annotation_moderation_test.py [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+
+    def __call__(self, permission, context):
+        return self._permissions[(permission, context)]
+
+
+class TestAnnotationModerationFormatter(object):
+    def test_preload_sets_flag_counts(self, db_session, flagged, unflagged, user):
+        formatter = AnnotationModerationFormatter(db_session,
+                                                  user,
+                                                  has_permission=None)
```
Comment (nickstenning) [inline tests/h/formatters/annotation_moderation_test.py]: `has_permission=None` is  presumably a convenience, in the sense that you just happen not to be calling any methods in this test that would explode because you haven't passed it in?

File: h/formatters/annotation_moderation.py (hunk 2 above)
Comment (fatbusinessman) [inline h/formatters/annotation_moderation.py]: Good point; added.

File: h/formatters/annotation_moderation.py (hunk 3 above)
Comment (fatbusinessman) [inline h/formatters/annotation_moderation.py]: Good point. I personally prefer that too; I was following the convention I’ve seen elsewhere in the codebase, but I’d be happy to go with this one instead.

Given this is code I’m moving around in #4502, I’m going to take your suggestion, but over in the other pull request.

File: h/formatters/annotation_moderation.py (hunk 4 above)
Comment (fatbusinessman) [inline h/formatters/annotation_moderation.py]: You’re right: it would. Again, I’ll address this over in #4502.

File: h/formatters/annotation_moderation.py (hunk 5 above)
Comment (fatbusinessman) [inline h/formatters/annotation_moderation.py]: Having looked at it again, I realised that this path will never actually get run, because if the user is `None`, they’ll fail the permission check in `format` and this method won’t get called. I’ve removed the check.

File: h/formatters/annotation_moderation.py (hunk 6 above)
Comment (fatbusinessman) [inline h/formatters/annotation_moderation.py]: Done.

File: tests/h/formatters/annotation_moderation_test.py (hunk 7 above)
Comment (fatbusinessman) [inline tests/h/formatters/annotation_moderation_test.py]: Correct. I’ve managed to eliminate most of these cases (except the one where the user is `None`) by extracting a fixture for the common case (of a user who has permission to see the `moderation` key).

**Comment by fatbusinessman** [comment]
Ok. I think I’ve addressed all the feedback (thanks, @nickstenning!) and rebased onto `master` to accommodate #4499. Give or take Jenkins having a bit of a sad at the moment, I hope this is ready for review again (and, give or take some `autosquash` magicks that I’ve deferred to make review easier, for merging).

**Comment by nickstenning** [comment]
LGTM pending a `rebase --autosquash`.
